### DISCUSSION TITLE
Generate student version for external files

### DISCUSSION THREAD

**Discussion by guyhoffman** [jupyter/nbgrader#1818] [Q&A] [answered]
Hi,

I'm using nbgrader for the first time this semester, and in my assignment, students both write code in the notebook, but also implement functions in external files. They fill out code segments between specifically structured comments, like this: 

```
    def transition(self):
        """Apply HMM transition model
        """
        ############################################################################
        # TODO: Add any needed initialization for the HMM class here.              #
        ############################################################################
        # YOUR CODE HERE

        raise NotImplementedError()

        ############################################################################
        #                             END OF YOUR CODE                             #
        ############################################################################
```

Is there any way to have the "Generate" function also parse and replace code segments in files outside the notebook? I know that it is copying the whole directory, so maybe there's a hook there? 

Right now I am using this somewhat hacky shell script to do the work, but I have to run this step separate and manage both versions of the file.

```
RELEASE_CODE='raise NotImplementedError()'
SUB_CMD='s/(\s*)(# YOUR CODE HERE)(\s|.)*?(#{76})/\1\2\n\1'$RELEASE_CODE'\n\1\4/g'

for file in mae6710/*.py(.); do 
    echo 'Converting '$file'...'
    perl -0777pe  $SUB_CMD $file > release/$(basename $file)
done
```

Thanks for this awesome tool!

**Comment by ranjiGT** [comment, marked as answer]
Hi @guyhoffman ,

you’re right that nbgrader will copy the whole directory, but it only actively processes *notebooks*. It doesn’t natively parse or transform external `.py` files in the same way it handles `# YOUR CODE HERE` regions inside notebook cells.

That said, you don’t need to rely on a separate shell script—there *is* a cleaner way using nbgrader’s extension points.

What you’re looking for is a **custom preprocessor** in the nbconvert pipeline used by `nbgrader generate_assignment`. With that, you can hook into the generation step and modify arbitrary files (including `.py`) before they’re written to the release directory.

The rough approach is:

* subclass `nbconvert.preprocessors.Preprocessor`
* in `preprocess`, locate your external files (they’re already copied into the temporary build directory at that point)
* apply your regex or transformation in Python
* overwrite the files in-place before the release is finalized

Then register it in your config, e.g.:

```python id="kqgqzb"
c = get_config()

c.GenerateAssignment.preprocessors = [
    'path.to.YourExternalFilePreprocessor',
]
```

Inside your preprocessor you can essentially port your current Perl logic to Python (`re.sub` with `re.DOTALL`), which makes the whole workflow self-contained in nbgrader.

A few additional notes:

* This keeps a **single source of truth** (no dual file management).
* It works with any spawner setup (Docker/Kubernetes) since it’s part of the build step.
* You can even generalize it (e.g., support multiple markers or different templates).

Unfortunately, there isn’t a built-in feature specifically for external file regions like `# YOUR CODE HERE`, so a preprocessor (or your current script) is the intended level of customization.

Hope that helps—and your Perl solution is actually pretty clever, just a bit outside the nbgrader pipeline 🙂
